[Feat] 일기 상세 화면(1-1) 구현

ParkSeongGeun opened the pull request:
### 📌 관련 이슈 번호 
- #36 

<br>

# 📘 작업 유형
 - [x] 화면 구성
 - [x] `VC, VM, View, Coordinator`

<br>

# 📙 작업 내역 (구현 내용 및 작업 내역을 기재합니다.)
- 1-1 일기 상세를 위한 `DiaryDetailView, ViewController, ViewModel, Coordinator`를 생성했습니다.

<br>
## 실행 화면

https://github.com/user-attachments/assets/4d3d1352-154f-4f0e-b2ca-16fbc6d205aa



### 📝 특이 사항 (Optional)
- 일기 상세의 경우 `ViewController`의 생성자에서 `parameter`로 `TimeStamp`가 넘어온다고 가정하였습니다. (DataStorage 방식을 적용)
    - 예로, 2024년 11월 21일에 해당되는 달력 내 날짜 클릭시 20241121에 해당되는 정보가 ViewController 생성자 파라미터로 들어옵니다.

- 이때 기존 `Coordinator` 프로토콜 내 `start()`로는 구현이 되지 않아, `DiaryDetailCoordinator`에서 `start(timeStamp: String)` 함수를 명시하여 구체 구현 객체에서 메서드를 생성해주었습니다.

- ScrollView의 스크롤바는 숨김처림을 진행했습니다.

- useCase를 금일 진행한 DataStorage의 형태로 만들어주었습니다. 일기 상세 화면 진입 시 DataStorage를 통해 Diary 객체가 반환되는 상황을 가정하여 UI를 구성하였습니다.

- RecordView 부분에서도 피드백 받은 단방향 Flow를 위한 수정을 진행하였습니다!

#### Q. navigationBar가 기존 애플 기본 제공용이 아닌, 커스텀한 디자인이 필요할 것 같다고 느꼈는데, 확인 후 논의도 필요할 것 같습니다!

ParkSeongGeun committed:
[Feat] #36 Create Coordinator, View, VC, VM

- 밑작업 진행

ParkSeongGeun committed:
Merge branch 'develop' into Feat/Detail/36

ParkSeongGeun committed:
[Feat] #36 Create DiaryDetailViewModel

ParkSeongGeun committed:
[Feat] #36 Create DiaryDetailViewController

ParkSeongGeun committed:
[Feat] #36 Create DiaryDetailView

ParkSeongGeun committed:
[Feat] #36 Create DiaryDetailCoordinator

ParkSeongGeun committed:
[Chore] #36 Create Constraints, Add TODO

futuremirae commented on Heim/Presentation/Presentation/Detail/DiaryDetail/View/DiaryDetailView.swift:
```diff
@@ -0,0 +1,207 @@
+
+private extension DiaryDetailView {
+  enum LayoutConstants {
+    static let dateLabelFontSize: CGFloat = 28
```
폰트사이즈의 경우 저번에 피그마에 나와있는데로 
컨벤션을 맞추기로 했던걸로 기억합니다~!

ParkSeongGeun replied:
앗 맞습니다. 수정하겠습니다!

futuremirae reviewed (approved):
빠르게 그리셨네요! 
성근님 고생 많으셨습니다 ~~

clxxrlove commented on Heim/Presentation/Presentation/Detail/DiaryDetail/View/DiaryDetailViewController.swift:
```diff
@@ -0,0 +1,106 @@
+          date: "2024년 10월 31일",
+          description: "성근님, 화가 나는 하루를 보내셨군요.",
+          content: """
+            이 지면에 실리는 글은 띄어쓰기 포함 700자 안팎이다. 초고를 쓰고 문서 통계창을 열어 글자수를 확인한 다음 분량을 조정한다. 
```
ㅋㅋㅋ 이거 약간 예전 스타크래프트 립버전 검색하면 나오는 결과같네요

clxxrlove commented on Heim/Presentation/Presentation/Detail/DiaryDetail/View/DiaryDetailViewController.swift:
```diff
@@ -0,0 +1,106 @@
+    
+    // TODO: state를 통해 configure를 사용할 예정
+    viewModel.$state
+      .receive(on: DispatchQueue.main)
```
P1. 다른 PR에도 리뷰했던 내용이네요.

ParkSeongGeun replied:
해당 부분은 
https://github.com/boostcampwm-2024/iOS05-Heim/pull/29#discussion_r1851329493

을 이유로 보류하겠습니다.

clxxrlove commented on Heim/Presentation/Presentation/Detail/DiaryDetail/ViewModel/DiaryDetailViewModel.swift:
```diff
@@ -0,0 +1,66 @@
+  // TODO: 기능 구현
+  func setUp() async {
+    state.date = timeStamp
+//    do {
+//      let diary = try await useCase.readDiary(timeStamp: timeStamp)
+//      state.date = timeStamp // 20241121만 추출하도록
+//      state.description = diary.emotion
+//      state.content = diary.Summary.text
+//    } catch {
+//      // TODO: Error Handling
+//    }
```
P4. 이런 느낌의 주석이 많이 보이는데요. 차라리 TODO를 것도 좋은 방법이지 않나 싶은 개인적인 생각입니다. 물론 나중에 사용하기 위해 주석을 두는 것도 맞지만, 이런 주석이 많으면 미완성 코드의 느낌이 나는 것도 사실인거 같아요.

ParkSeongGeun replied:
맞습니다. 제 개인적인 욕심으로 인해 발생한 문제 같습니다.

추후 useCase만 merge되면 바로 사용하려고 했었는데, 가독성을 저해시키는 요인이 되는 거 같습니다.

말씀 감사합니다 :)

clxxrlove commented on Heim/Presentation/Presentation/Detail/DiaryDetail/View/DiaryDetailView.swift:
```diff
@@ -0,0 +1,207 @@
+  }
+  
+  // MARK: - Action Methods
+  @objc func musicRecomendationButtonTapped() {
+    delegate?.buttonDidTap(self, .musicRecomendation)
+  }
+  
+  @objc func heimReplyButtonTapped() {
+    delegate?.buttonDidTap(self, .heimReply)
+  }
+  
+  @objc func replayVoiceButtonTapped() {
+    delegate?.buttonDidTap(self, .replayVoice)
+  }
```
이거 되게 좋은 방법이네요 ? 배워갑니다 !

ParkSeongGeun replied:
미리 enum으로 정의한 거를 쓰는 것, delegate로 동작을 넘기는 것으로 구현했습니다!

clxxrlove commented on Heim/Presentation/Presentation/Detail/DiaryDetail/View/DiaryDetailView.swift:
```diff
@@ -0,0 +1,207 @@
+  
+  private let characterImageView: UIImageView = {
+    let imageView = UIImageView()
+    imageView.image = UIImage(named: "splashRabbit")
```
P5. 저도 고민되는 부분인데요. splashRabbit이라거나 Asset에 추가한 이미지들을 이렇게 상수로 꺼냈을 때 분명히 실수가 있을 수도 있는 부분이라고 생각하는데, 어떤 좋은 방법이 있을지 생각해봐도 좋을 거 같아요. 다음에 같이 이야기 해볼까요 ?

ParkSeongGeun replied:
좋습니다! 현재는 Asset Catalog에 잇는 요소를 .recordRabbit과 같이 꺼내오는 방식으로 수정 후 사용할 예정입니다.

clxxrlove commented on Heim/Presentation/Presentation/Detail/DiaryDetail/View/DiaryDetailView.swift:
```diff
@@ -0,0 +1,207 @@
+  weak var delegate: DiaryDetailViewDelegate?
+  private let textArea = CommonTextAreaView()
+  
+  private lazy var scrollView: UIScrollView = {
```
여기서 `lazy` 키워드를 사용한 이유가 어떤걸까요 ? 자세한 설명을 들어보고 싶어요. 저도 많이 헷갈리는 부분이라서요.

ParkSeongGeun replied:
SwiftUI에서 scrollview를 사용시 내부 요소들을 LazyVStack을 사용했던 기억이 있어, scrollView에도 lazy var가 들어갈 거라 생각만 하고 넣었었습니다.

- ~~사실 사용할 이유가 없습니다.~~
- 수정하였습니다.

clxxrlove reviewed (approved):
아쉬운 부분이 있었다면, 코드 주석이 많았던 부분이 조금 아쉬웠어요. 제 개인적으로요.
하지만 그 외에는 정말 깔끔하게 잘 작성해주신거 같아요. 배워야겠다 싶은 내용들도 많이 있었고요.
정말 수고 많으셨습니다 :) 작업량이 엄청나시군요

futuremirae commented on Heim/Presentation/Presentation/Detail/DiaryDetail/View/DiaryDetailView.swift:
```diff
@@ -0,0 +1,207 @@
+    return imageView
+  }()
+  
+  private lazy var buttonStackView: UIStackView = {
```
let으로 안하고 lazy var 사용하신 이유가 있나요?!

ParkSeongGeun replied:
아 관성... 코딩이 바로 수정하겠습니다.
lazy var가 아래서 부터 들어가있는데, 들어갈 이유가 없습니다 현재는

ParkSeongGeun committed:
Merge branch 'develop' into Feat/Detail/36

ParkSeongGeun committed:
[Fix] #36 PR Review

- 불필요한 lazy var -> let으로 교체
- LayoutContstants 수정

ParkSeongGeun committed:
Merge branch 'develop' into Feat/Detail/36

ParkSeongGeun committed:
[Fix] #36 PR Review

- action, state를 활용한 데이터 단방향 흐름만 가능하도록 수정
- timeStamp -> Diary 정보가 VC 생성자 파라미터로 들어오도록 수정
- TODO: Diary Entity에 날짜 정보 포함해야함.

ParkSeongGeun committed:
[Fix] #36 PR Review
